### ISSUE TITLE
AOTAutograd should put keep mutations in the graph during training

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#109240]
Tracking issue for future work in aot autograd.

Today, if you compile a model that has input mutations, and you're compiling an inference graph (because e.g. you're running your model under `no_grad()`), then AOTAutograd will do the following:

(1) functionalize the graph; `x.mul_(2)` becomes `x_updated = x.mul(2)`
(2) detect that there were input mutations, which it represents as extra `x.copy_(x_updated)` calls
(3) Put those copy_() nodes directly in the graph so that inductor can fuse them

However, this is not true in the training case. Instead, AOTAutograd will:

(1) return `x_updated` as an additional graph output
(2) run an opaque epilogue (in eager mode), that manually performs the mutation: `x.copy_(x_updated)`

Historically, getting input mutations into the graph for inference was important because of optimizers: optimizers are (usually) run under no_grad so we get an inference graphs, and captured graphs of optimizers have a large number of input mutations, due to each parameter (a graph input) being mutated. So we made sure that when you compile optimizer code as an inference graph, we keep the input mutations in the graph for inductor to fuse.

It would still be useful to keep input mutations in the graph for the training case too, though. One example is batch norm: today, inductor can't actually see and speed up the buffer mutations that occur from batch norm during training, because they are not kept in the graph.

The main difficult in getting this to work is that we need to make sure that when AOTAutograd invokes the partitioner to partition out a separate fw/bw graph, the "correct" thing needs to happen (if there are copy_() nodes in the joint graph due to input mutations, those nodes should always show up at the very end of the partitioned forward graph).

cc @ezyang @msaroufim @wconstab @anijain2305 @zou3519

**Comment by luccafong** [comment]
Add some example code to reproduce this issue
```python
from torch import nn
class InplaceModel(nn.Module):
    def __init__(self, dim=30):
        super().__init__()
        self.w = torch.nn.Parameter(torch.randn(dim, dim, device="cuda"))
        l = torch.rand(dim, dim, dim, device="cuda")
        self.register_buffer('l', l)
    def forward(self, x):
        # x = torch.randn(256, 256, 256)
        x = x.add_(0.01)
        x = x.add_(0.01)
        x = x.add_(0.01)
        x = x.add_(0.01)
        x = x.add_(0.01)
        x = x.add_(0.01)
        x = x.add_(0.01)
        self.l.add_(x)
        x = self.l * x
        self.l.add_(x)
        x = self.l * x
        # x = self.w * x
        x = self.w * x
        x = self.w * x
        return x
from torch import nn
class Model(nn.Module):
    def __init__(self, dim=30):
        super().__init__()
        self.w = torch.nn.Parameter(torch.randn(dim, dim, device="cuda"))
    def forward(self, x):
        # x = torch.randn(256, 256, 256)
        x = x.add(0.01)
        x = x.add(0.01)
        x = x.add(0.01)
        x = x.add(0.01)
        x = x.add(0.01)
        x = x.add(0.01)
        x = x.add(0.01)
        x = self.w * x
        x = self.w * x
        x = self.w * x
        x = self.w * x
        return x
import triton
def benchmark_dynamo(model_class, dim):
    torch.manual_seed(0)
    print(model_class)
    x1 = torch.rand(dim, dim, dim).cuda()
    x2 = torch.rand(dim, dim, dim).cuda()
    model = model_class(dim).cuda()
    dynamo_model = torch._dynamo.optimize("inductor")(model)
    optimizer_1 = torch.optim.SGD(model.parameters(), lr=0.001, momentum=0.9)
    optimizer_2 = torch.optim.SGD(dynamo_model.parameters(), lr=0.001, momentum=0.9)
    def train_eager():
        optimizer_1.zero_grad()
        y = model(x1)
        y.sum().backward()
        optimizer_1.step()
    def train_dynamo():
        optimizer_2.zero_grad()
        y = dynamo_model(x2)
        y.sum().backward()
        optimizer_2.step()
    eager_latency = triton.testing.do_bench(train_eager, 50, 100)
    dynamo_latency = triton.testing.do_bench(train_dynamo, 50, 100)
    ratio = (dynamo_latency / eager_latency - 1) * 100
    result =  (f"Eager latency: {eager_latency} ms, dynamo latency: {dynamo_latency} ms, ratio: {ratio} %")
    print(result)
    return result
def benchmark_outplace_vs_inplace(model_class, model_class_inplace, dim):
    print(f"dim={dim}")
    benchmark_dynamo(model_class, dim)
    benchmark_dynamo(model_class_inplace, dim)
benchmark_outplace_vs_inplace(Model, InplaceModel, 32)
benchmark_outplace_vs_inplace(Model, InplaceModel, 64)
benchmark_outplace_vs_inplace(Model, InplaceModel, 128)
benchmark_outplace_vs_inplace(Model, InplaceModel, 256)
benchmark_outplace_vs_inplace(Model, InplaceModel, 512)
```

output shows dynamo with inplace operators perform worse
```
dim=32
<class '__main__.Model'>
Eager latency: 0.6409690380096436 ms, dynamo latency: 0.5889919996261597 ms, ratio: -8.109134030074927 %
<class '__main__.InplaceModel'>
Eager latency: 0.5481123328208923 ms, dynamo latency: 0.7541775107383728 ms, ratio: 37.59542808623808 %
dim=64
<class '__main__.Model'>
Eager latency: 0.6558034420013428 ms, dynamo latency: 0.6327101588249207 ms, ratio: -3.5213726701322834 %
<class '__main__.InplaceModel'>
Eager latency: 0.5504335761070251 ms, dynamo latency: 0.7018740773200989 ms, ratio: 27.51294757201148 %
dim=128
<class '__main__.Model'>
Eager latency: 0.6792190074920654 ms, dynamo latency: 0.6047821640968323 ms, ratio: -10.95918143841148 %
<class '__main__.InplaceModel'>
Eager latency: 0.5461625456809998 ms, dynamo latency: 0.6858663558959961 ms, ratio: 25.579163441316233 %
dim=256
<class '__main__.Model'>
Eager latency: 2.4841315746307373 ms, dynamo latency: 0.6217321753501892 ms, ratio: -74.97185005417401 %
<class '__main__.InplaceModel'>
Eager latency: 2.171820878982544 ms, dynamo latency: 0.9222127795219421 ms, ratio: -57.537346267985924 %
dim=512
<class '__main__.Model'>
Eager latency: 18.227025985717773 ms, dynamo latency: 2.8842198848724365 ms, ratio: -84.17613555205091 %
<class '__main__.InplaceModel'>
Eager latency: 16.115732192993164 ms, dynamo latency: 6.45711612701416 ms, ratio: -59.93284047111679 %
```

**Comment by jon-chuang** [comment]
I was able to reproduce that the in-place modification to the `nn.Module` resulted in slowdown.

<details>

![image](https://github.com/pytorch/pytorch/assets/9093549/9ad3da2c-31d0-4e40-a59a-c5348bd7ed11)



</details>

@bdhirsh if you are not actively working on this I might take a look since I want to read about AOTAutograd.

**Comment by bdhirsh** [comment]
@jon-chuang sure - I'm not actively working on this, but I can try to advise.

One caveat about your script - I agree that it shows slowdowns in torch.compile, but it doesn't actually prove that "moving the copy nodes inside the graph" is exactly what would fix the slowdowns. A better experiment would probably be to run an experiment with compile under the pytorch profiler, and witness that a significant portion of the runtime is from us dispatching directly to `copy_()` kernels after the compiled graph has run.

**Comment by bdhirsh** [comment]
Another caveat on this issue: I think we probably want to limit this issue for now to **buffers** and model inputs that do **not** require gradients (this is much more common anyway, since e.g. batchnorm parameters do not require gradient).

Keeping input mutations on inputs that do require gradients is more complicated, since we the autograd engine to know about the input mutation, so we can't just blindly hide it in the graph.

**Comment by jon-chuang** [comment]
Oh, I see, in the batch norm case with `running_mean / running_var`, I guess it is similar to "cumulative layer norm" in the sense that it accumulates the statistics across mini-batches, and in this case, these do not require gradients.

> A better experiment would probably be to run an experiment with compile under the pytorch profiler, and witness that a significant portion of the runtime is from us dispatching directly to copy_() kernels after the compiled graph has run.

I think that as long as we can show a speedup on batchnorm, we ought to be quite happy for the purpose of step 1 (buffers / no gradients).

**Comment by jon-chuang** [comment]
I think the part that is a little confusing to me is:
> if there are copy_() nodes in the joint graph due to input mutations, those nodes should always show up at the very end of the partitioned forward graph

To my understanding, if we drop into SSA world, this copy is equivalent to assignment to var. However, the actual representation printed in the fwd graph is equivalent to:
```python
updated_input_1 = op input_1
...
copy = aten.copy.default(updated_input)
... # no further reference to copy
```

In this case, the input var was not used further in the fwd graph after the mutation. I find this rather confusing. Shouldn't the `GraphModule` be more like:

```python
updated_input_1 = op(input_1)
...
input_1 = aten.copy.default(updated_input_1)
```

### Joint Graph Case
What we actually want in the joint graph is something like:

```python
updated_input_1 = op(input_1)
...
input_1 = aten.copy.default(updated_input)
...
loss = ...
# bwd graph
...
%10 = bwd_op(updated_input_1, weights)

```
Then inductor can bufferize all references to `updated_input_1` to `input_1`.

Is this analysis correct?

**Comment by jon-chuang** [comment]
> A better experiment would probably be to run an experiment with compile under the pytorch profiler, and witness that a significant portion of the runtime is from us dispatching directly to copy_() kernels after the compiled graph has run.

My worry is that in batchnorm case the actual effect would be very small, as the `statistics space is << hidden dim space`. Still, the eager-mode allocation + copy may very well destroy performance. This is due possibly needing to call synchronize before running eager-mode code?

**Comment by jon-chuang** [comment]
>  think we probably want to limit this issue for now to buffers and model inputs that do not require gradients 

For this, the specific recipe would be to use eager mode copy in the case where there are gradients required, and to inline the copy into the graph when no gradients are required. 

In terms of the old code path we can treat the inlined copy input vars as not being mutated (del from `mutated_inputs`).

> Keeping input mutations on inputs that do require gradients is more complicated, since we the autograd engine to know about the input mutation, so we can't just blindly hide it in the graph.

I actually have a feeling this would still be fine - the mutations will be functionalized.

**Comment by jon-chuang** [comment]
Btw @bdhirsh, to double confirm, the invariant that we try to respect that the graph as a whole should not mutate any input buffers unless explicitly specified by the user code. That's because we cannot assume that the graph represents the full program the user is executing, and inductor/AOTAutograd cannot see "outside" the graph to determine if the buffer or var would be used elsewhere in the user code, correct?

**Comment by jon-chuang** [comment]
> if there are copy_() nodes in the joint graph due to input mutations, those nodes should always show up at the very end of the partitioned forward graph

I'm not sure how well this would play with `min_cut_rematerialization_partitioner`. I could try to enforce that copy always has edge with infinite weight from source. Hence, it will never get repartitioned into bwd graph. (Or otherwise forbid the rematerialization of this node).

Note however that your requested condition is too strict - it is not necessary for the copy to appear at the end of the fwd graph. A materialized downstream node that utilizes the copy might be used instead. It is up to the partitioned in this case.

**Comment by bdhirsh** [comment]
> Oh, I see, in the batch norm case with running_mean / running_var, I guess it is similar to "cumulative layer norm" in the sense that it accumulates the statistics across mini-batches, and these do not require gradients.

yep that's right

> Right now, inductor will copy out the mutations into a new buffer, which is used both as inputs in the backwards pass in the compiled graph and copied out into the original buffer in eager mode. However, this increases the number of buffers required, whereas if the original buffer were mutatated we would avoid the extra allocation/copy and it can directly be used as the input to backwards pass operators in calculating backprop.

Yep, this is mostly right. For "making the mutated forward input a direct input to the backward pass": I don't think this will be easy to do. That is also pretty independent from the partitioner-related problems mentioned in this issue, so I think it's fine to only handle non-grad-requiring inputs for now.


> Then inductor can bufferize all references to copy_1 to input_1... Is this description correct?

In the joint graph, I'd expect something like this:
```
def original_user_fn(x)
    x.add_(3)
    return torch.mul(x, 2)

def joint_graph(self, primals, tangents):
    primals_1 = fx_pytree.tree_flatten_spec([primals, tangents], self._in_spec)
    add = torch.ops.aten.add.Tensor(primals_1, 3)
    mul = torch.ops.aten.mul.Tensor(add, 2)
    # Note: the copy_() output in the graph is unused, and does not show up
    # anywhere in the backward graph
    # The assumption is that the partitioner will ensure that the copy_() shows up at the end
    # of the partitioned forward
    copy_ = torch.ops.aten.copy_.default(primals_1, add)
    mul_1 = torch.ops.aten.mul.Tensor(tangents_1, 2)
    return pytree.tree_unflatten([mul, mul_1], self._out_spec)
```

**Comment by bdhirsh** [comment]
> I'm not sure how well this would play with min_cut_rematerialization_partitioner. I could try to enforce that copy always has edge with infinite weight from source. Hence, it will never get repartitioned into bwd graph.

I'm not exactly sure how we want to do this - your approach might be reasonable. I was also hoping that the partitioner might notice that the output of the `copy_()` node is not used in the backward at all, and it is therefore pointless for it to move it into the backward in the first place.

> Note however that your requested condition is too strict - it is not necessary for the copy to appear at the end of the fwd graph. A materialized downstream node that utilizes the copy might be used in the bwd instead. It is up to the partitioner in this case. We are simply concerned with making it possible to be reused.

Hmm I don't think this is true. I expect the output of the `copy_(inp, inp_updated)` node to **never** be used in the backward graph. If one of the inputs to the forward is re-used in the backward graph, it should be fine for the backward graph to directly use `inp`, and rely on the fact that the original input that we're saving for backward will be mutated by the `copy_()` at the end of the forward. On the other hand, we **do** need to make sure that the input mutation doesn't show up until after the last usage sight of `original_input` in the forward. It might be simplest though to just always make the partitioner put the `copy_()` calls at the end though.

Although hmm, one problem you might run into is that for this example:
```
# lets say x does not require gradients but y does require gradients
def f(x, y):
    x.mul_(2)  # mutate an input
    return x.sin() + y  # this saves x_updated for backward
```

In the user code above, we will have mutated first, and **then** saved for backward. We'll need a test to be sure, but I think instead in AOTAutograd we might end up performing the save_for_backward first, and **then** the `input.copy_(input_updated)`.

**Comment by bdhirsh** [comment]
Hey @jon-chuang sorry for the whiplash - someone from the pytorch team is actually going to work on this issue for onboarding (there are precious few onboard-friendly AOTAutograd tasks :( )

**Comment by jon-chuang** [comment]
@bdhirsh , no worries, I can help review, do @ me when it's ready :)

**Comment by jon-chuang** [comment]
> it should be fine for the backward graph to directly use inp, and rely on the fact that the original input that we're saving for backward will be mutated by the copy_() at the end of the forward.

Well yes, I guess this makes sense - `copy_1` will either be used in the bwd graph (as a primal input) or not. 

However, do note that _if `copy_1` is mutated_ in further fwd graphs (assuming there are graph breaks) we can no longer guarantee that it can be used as `inp` or `primal` to the bwd graph. We must then indeed issue a copy.

That's because the backprop would depend on the value of the buffer _at a particular cut_ in the forward graphs.


---

More generally:

I am unfamiliar, but it seems that given fwd or bwd graphs `g1, g2`, the logic to copy the buffer from output of g1 as the input to g2 `iff (buffer in g1.outputs and is_used_elsewhere(buffer, g2)) and (buffer in g2.input_mut)` should already be in existence. This seems to be once again related to `move` semantics of the buffers.

Hence, I am unclear if the optimizations that one desires can even be performed reasonably, since `is_used_anywhere` is impossible to know (the buffer may be used in some eagerly-executed graph break, for instance). It seems that _as long as the buffer is owned by the compilation pass_, then the reasoning would be simplified (alias analysis).

**Comment by jon-chuang** [comment]
> I'm not exactly sure how we want to do this - your approach might be reasonable. I was also hoping that the partitioner might notice that the output of the copy_() node is not used in the backward at all, and it is therefore pointless for it to move it into the backward in the first place.

Actually, thinking about the whole point of the min-cut partitioner (see [pytorch form post](https://dev-discuss.pytorch.org/t/min-cut-optimal-recomputation-i-e-activation-checkpointing-with-aotautograd/467)), we might actually want to move the mutations _into the bwd graph_.

> Yep, this is mostly right. For "making the mutated forward input a direct input to the backward pass": I don't think this will be easy to do. That is also pretty independent from the partitioner-related problems mentioned in this issue, so I think it's fine to only handle non-grad-requiring inputs for now.

I think you mistake what I am saying - what I mean is that the buffers resulting from mutations may participate in the bwd computation, even if they require no gradients themselves. As a matter of fact, it may not even be necessary to mutate the buffer in the forward pass, since we might decide to mutate it in the backward... 

In fact, _if it is mutated in the bwd, there is no need to mutate it in the fwd_.

**Comment by jgong5** [comment]
> My worry is that in batchnorm case the actual effect would be very small, as the `statistics space is << hidden dim space`. So allocation + read/write overhead should be relatively small. Still, the eager-mode allocation + copy may very well destroy performance. This is due possibly needing to call `torch.cuda.synchronize()` before running eager-mode code?

Yep, it is primarily due to the overhead in kernel launch and scheduling for these small kernels. We do see non-trivial overhead in RN50 training on our Habana accelerators.

**Comment by bdhirsh** [comment]
I talked to @zou3519 about this issue, he made two additional point:

(1) For the training case, we need to make sure to call `mark_dirty()` for every tensor where we move the `copy_()` inside the graph, so autograd knows to update the version counter

(2) It will be pretty hard to properly handle an input mutation case like `inp_not_requiring_grad.mul_(tensor_requiring_grad)`, since this mutation will cause the input to start requiring grad. We'll probably need to carefully detect this case, and not try to put the input mutation in the graph

**Comment by Chillee** [comment]
An argument that this may have somewhat significant perf improvements is that https://github.com/pytorch/pytorch/pull/110721 had a pretty significant impact.
